src/sync.jl

```diff
@@ -0,0 +1,65 @@
+module Sync
+    global mutex = ReentrantLock()
+    global sync_level = repeat([ 0 ], Threads.nthreads())
+
+    @inline is_locked() = sync_level[Threads.threadid()] > 0
+
+    @inline function lock()
+        tid = Threads.threadid()
+        if sync_level[tid] == 0
+            Base.lock(mutex)
+        end
+        sync_level[tid] += 1
+    end
+
+    @inline function unlock()
+        tid = Threads.threadid()
+        sync_level[tid] -= 1
+        if sync_level[tid] == 0
+            Base.unlock(mutex)
+        end
+    end
+
+    @inline function check_lock()
+        assert(is_locked())
+    end
+end
+
+macro sync(expr)
+    if Threads.nthreads() > 1
+        quote
+            try
+                Sync.lock()
+                $(esc(expr))
```

Comment:
Gah, I really can not (yet?) deal with this "value of last expression is automatically returned"... I was about to comment that something like what is done in `sync_noexcept` is needed here, too, but I gather that in a regular execution, this line is considered to be the "last" expression, and its value is returned. Right?

A quick test in Julia seems to confirm that.

Reply:
It's something that Julia has borrowed from Ruby (I think) in that every statement also forms an expression. For a block statement, the value of the last statement/expression is the value of the block statement as a whole; ditto for try/catch.

It's actually pretty nice for macros in that you can treat most statements and expressions the same.